development/tools/cmd/oomfinder/main.go

```diff
@@ -5,118 +5,154 @@ import (
+			case err := <-oomErrCh:
+				log.WithError(err).Error("got containerd event on errors channel")
 				return
 			}
-			log.WithFields(log.Fields{"msg": "got containerd events errors channel event"}).Errorf("%+v", cherr)
-		}
-	}()
-	go func() {
-		defer wg.Done()
-		for {
-			event, ok := <-events
-			if !ok {
-				log.Errorf("failed read containerd events channel")
-				return
+			if e != nil {
+				if e.Event != nil {
+					log.Debug("Got containerd oom event.")
+					v, err := typeurl.UnmarshalAny(e.Event)
+					if err != nil {
+						log.WithError(err).Warn("cannot unmarshal an event from Any")
+						continue
+					}
+					log.Debugf("got oom event for containerID: %s\n", v.(*events2.TaskOOM).ContainerID)
+					ctxWithNamespace := namespaces.WithNamespace(ctx, "k8s.io")
+					container, err := client.ContainerService().Get(ctxWithNamespace, v.(*events2.TaskOOM).ContainerID)
+					if err != nil {
+						log.WithError(err).Error("Failed read container details.")
+						continue
+					}
+					log.Debugf("Image: %s, Name: %s, Name: %s, Namespace: %s", container.Image, container.Labels["io.kubernetes.container.name"], container.Labels["io.kubernetes.pod.name"], container.Labels["io.kubernetes.pod.namespace"])
+					_, err = fmt.Fprintf(os.Stdout, "OOM event received, time: %s, namespace: %s, pod: %s, container: %s, image: %s\n",
+						e.Timestamp,
+						container.Labels["io.kubernetes.pod.namespace"],
+						container.Labels["io.kubernetes.pod.name"],
+						container.Labels["io.kubernetes.container.name"],
+						container.Image,
+					)
+					if err != nil {
+						log.WithError(err).Error("cannot print EventInfo")
+						continue
+					}
+				}
 			}
-			fmt.Printf("%+v", event)
 		}
 	}()
-	//TODO: check what data is in containerd oom event and if following code is needed.
-	oom := namespaces.WithNamespace(context.Background(), "oom")
-	allContainers, err := client.Containers(oom)
-	if err != nil {
-		log.WithFields(log.Fields{"msg": "failed get containers"}).Fatalf("error: %s", err)
-	}
-	for container := range allContainers {
-		fmt.Fprintf(os.Stdout, "%+v", container)
-	}
 }
+
 func main() {
-	// wait group to allow goroutines listen on channels
+	flag.Parse()
+	if *debug {
+		log.SetLevel(log.DebugLevel)
+	}
+	// Wait group to allow goroutines listen on channels.
 	var wg sync.WaitGroup
-	// check docker daemon socket exists
+	var socketFlag atomic.Value
+	// Check docker daemon socket exists.
 	if _, err := os.Stat("/var/run/docker.sock"); err == nil {
-		// create docker client with unix socket
+		socketFlag.Store(1)
+		// Create docker client with unix socket.
 		client, err := dockerclient.NewClient("unix:///var/run/docker.sock")
 		if err != nil {
-			log.WithFields(log.Fields{"msg": "failed create docker client"}).Fatalf("error: %s", err)
+			log.WithError(err).Fatalf("failed create docker client")
 		}
 		wg.Add(1)
-		// listen for oom events
+		// Listen for oom events.
 		dockerOOMListener(client, &wg)
```

Comment:
I have a small concern regarding this function. If this workload is added to the workgroup then shouldn't this entire function be run in the separate thread? I mean run this function in separate thread and remove `go func()` call inside it. 
```go
go dockerOOMListener(client, &wg)
```

Reply:
I would stick with current place for starting goroutine because it runs a minimal required part of application in separate threads. Code which doesn't require to run in parrallel is outside this additionall routines.